### PR TITLE
slightly improve jemalloc allocator API header

### PR DESCRIPTION
**Pull request by jsteemann**
Fix a few typos and avoid a potential nullptr dereference.

### CODE REVIEW

File: include/rocksdb/memory_allocator.h [hunk 1]
```diff
@@ -45,31 +45,31 @@ struct JemallocAllocatorOptions {
...
-// The JemallocNodumpAllocator creates a delicated jemalloc arena, and all
-// allocations of the JemallocNodumpAllocator is through the same arena.
+// The JemallocNodumpAllocator creates a dedicated jemalloc arena, and all
+// allocations of the JemallocNodumpAllocator are through the same arena.
 // The memory allocator hooks memory allocation of the arena, and call
-// madvice() with MADV_DONTDUMP flag to exclude the piece of memory from
-// core dump. Side benefit of using single arena would be reduce of jemalloc
+// madvise() with MADV_DONTDUMP flag to exclude the piece of memory from
+// core dump. Side benefit of using single arena would be reduction of jemalloc
 // metadata for some workload.
```
Comment (mrambacher) [inline include/rocksdb/memory_allocator.h:67]: some workloads"

File: include/rocksdb/memory_allocator.h [hunk 2]
```diff
@@ -45,31 +45,31 @@ struct JemallocAllocatorOptions {
...
 //
 // To mitigate mutex contention for using one single arena, jemalloc tcache
 // (thread-local cache) is enabled to cache unused allocations for future use.
-// The tcache normally incur 0.5M extra memory usage per-thread. The usage
-// can be reduce by limitting allocation sizes to cache.
+// The tcache normally incurs 0.5M extra memory usage per-thread. The usage
+// can be reduced by limitting allocation sizes to cache.
```
Comment (mrambacher) [inline include/rocksdb/memory_allocator.h:72]: "limiting", not "limitting".

File: include/rocksdb/memory_allocator.h [hunk 3]
```diff
@@ -45,31 +45,31 @@ struct JemallocAllocatorOptions {
...
+// MADV_DONTDUMP through madvise to exclude cache items from core dump.
 // Applications can use the allocator with block cache to exclude block cache
 // usage from core dump.
 //
 // Implementation details:
-// The JemallocNodumpAllocator creates a delicated jemalloc arena, and all
-// allocations of the JemallocNodumpAllocator is through the same arena.
+// The JemallocNodumpAllocator creates a dedicated jemalloc arena, and all
+// allocations of the JemallocNodumpAllocator are through the same arena.
 // The memory allocator hooks memory allocation of the arena, and call
```
Comment (mrambacher) [inline include/rocksdb/memory_allocator.h:64]: "and calls"

File: include/rocksdb/memory_allocator.h [hunk 4]
```diff
@@ -45,31 +45,31 @@ struct JemallocAllocatorOptions {
...
+  // When used with block cache, it is recommended to set it to block_size/4.
   size_t tcache_size_lower_bound = 1024;
 
   // Upper bound of allocation size to use tcache, if limit_tcache_size=true.
-  // When used with block cache, it is recommneded to set it to block_size.
+  // When used with block cache, it is recommended to set it to block_size.
   size_t tcache_size_upper_bound = 16 * 1024;
 };
 
 // Generate memory allocators which allocates through Jemalloc and utilize
```
Comment (mrambacher) [inline include/rocksdb/memory_allocator.h:56]: "Generate memory allocator"

**Review by mrambacher — COMMENTED** [review]
LGTM, with the few minor comments.

File: include/rocksdb/memory_allocator.h (hunk 4 above)
Comment (jsteemann) [inline include/rocksdb/memory_allocator.h:56]:
```suggestion
// Generate memory allocator which allocates through Jemalloc and utilize
```

File: include/rocksdb/memory_allocator.h [hunk 5]
```diff
@@ -45,31 +45,31 @@ struct JemallocAllocatorOptions {
...
-// The JemallocNodumpAllocator creates a delicated jemalloc arena, and all
-// allocations of the JemallocNodumpAllocator is through the same arena.
+// The JemallocNodumpAllocator creates a dedicated jemalloc arena, and all
+// allocations of the JemallocNodumpAllocator are through the same arena.
 // The memory allocator hooks memory allocation of the arena, and call
-// madvice() with MADV_DONTDUMP flag to exclude the piece of memory from
-// core dump. Side benefit of using single arena would be reduce of jemalloc
+// madvise() with MADV_DONTDUMP flag to exclude the piece of memory from
+// core dump. Side benefit of using single arena would be reduction of jemalloc
 // metadata for some workload.
```
Comment (jsteemann) [inline include/rocksdb/memory_allocator.h:67]:
```suggestion
// metadata for some workloads.
```

File: include/rocksdb/memory_allocator.h [hunk 6]
```diff
@@ -45,31 +45,31 @@ struct JemallocAllocatorOptions {
...
+// madvise() with MADV_DONTDUMP flag to exclude the piece of memory from
+// core dump. Side benefit of using single arena would be reduction of jemalloc
+// metadata for some workloads.
 //
 // To mitigate mutex contention for using one single arena, jemalloc tcache
 // (thread-local cache) is enabled to cache unused allocations for future use.
-// The tcache normally incur 0.5M extra memory usage per-thread. The usage
-// can be reduce by limitting allocation sizes to cache.
+// The tcache normally incurs 0.5M extra memory usage per-thread. The usage
+// can be reduced by limitting allocation sizes to cache.
```
Comment (jsteemann) [inline include/rocksdb/memory_allocator.h:72]:
```suggestion
// metadata for some workloads.
```
```suggestion
// can be reduced by limiting allocation sizes to cache.
```

File: include/rocksdb/memory_allocator.h [hunk 7]
```diff
@@ -45,31 +45,31 @@ struct JemallocAllocatorOptions {
...
+// MADV_DONTDUMP through madvise to exclude cache items from core dump.
 // Applications can use the allocator with block cache to exclude block cache
 // usage from core dump.
 //
 // Implementation details:
-// The JemallocNodumpAllocator creates a delicated jemalloc arena, and all
-// allocations of the JemallocNodumpAllocator is through the same arena.
+// The JemallocNodumpAllocator creates a dedicated jemalloc arena, and all
+// allocations of the JemallocNodumpAllocator are through the same arena.
 // The memory allocator hooks memory allocation of the arena, and call
```
Comment (jsteemann) [inline include/rocksdb/memory_allocator.h:64]:
```suggestion
// The memory allocator hooks memory allocation of the arena, and calls
```

**Comment by jsteemann** [comment]
@mrambacher : thanks for the review. I incorporated your suggestions into the PR.

File: memory/jemalloc_nodump_allocator.cc [hunk 8]
```diff
@@ -132,7 +132,9 @@ size_t JemallocNodumpAllocator::UsableSize(void* p,
 Status NewJemallocNodumpAllocator(
     JemallocAllocatorOptions& options,
     std::shared_ptr<MemoryAllocator>* memory_allocator) {
-  *memory_allocator = nullptr;
+  if (memory_allocator != nullptr) {
```
Comment (riversand963) [inline memory/jemalloc_nodump_allocator.cc:135]: How about checking `memory_allocator` with `nullptr` just once by combining this with another check below.
```
if (memory_allocator) {
  *memory_allocator = nullptr;
} else {
  return Status::InvalidArgument("memory_allocator must be non-null.");
}
```

**Review by riversand963 — COMMENTED** [review]
Thanks @jsteemann for the improvement.

File: memory/jemalloc_nodump_allocator.cc (hunk 8 above)
Comment (jsteemann) [reply memory/jemalloc_nodump_allocator.cc]: @riversand963 : Ok, I adjusted the PR once again.

**Review by riversand963 — APPROVED** [review]

**Pull request closed without merging**